### DISCUSSION TITLE
Adding support for Wasp completions in `zsh`

### DISCUSSION THREAD

**Discussion by infomiho** [wasp-lang/wasp#964] [Archived]
We currently have `bash` completions support with `wasp completion` and `wasp completion:generate > <your-chosen-directory>/wasp-completion`. https://github.com/wasp-lang/wasp/blob/2ed921a44273e41a3f287c6c56df1cab10805b10/waspc/cli/src/Wasp/Cli/Command/BashCompletion.hs#L16

These completions are `bash` specific and don't work with `zsh` (default shell for our MacOS users). The user needs to add the following to their `.zshrc` to add support for `bash`-like completion:

```
autoload -U compinit && compinit -u
autoload bashcompinit && bashcompinit
```

We want to make our `completion` command aware that there are two different shells and help users with `zsh` to setup Wasp completions. The command should allow user to select which shell he uses and get the proper info how to setup completions.

For example, `wasp completion zsh` would instruct user with `zsh` how to setup completions.

We could take two routes:
1. Piggy-back on `bash` completions and instruct user how to enable it for `zsh` (quick and dirty)
2. Write our own native `zsh` completions code (proper way to do it [details](https://blog.mads-hartmann.com/2017/08/06/writing-zsh-completion-scripts.html))

**Comment by infomiho** [comment]
I've go this working for me (it's pretty basic, since I'm new to zsh completions).

file: `~/.zshrc`
```bash
fpath=($fpath ~/.completions)
```

file: `~/.completions/_wasp`
```bash
#compdef _wasp wasp

function _wasp {
    local line

    _arguments -C \
        "1: :(new version waspls completion start db clean build telemetry deps dockerfile info)" \
        "*::arg:->args"

    case $line[1] in
        new)
            _wasp_new
        ;;
        db)
            _wasp_db
        ;;
    esac
}

function _wasp_new {
    _arguments \
        "1:project name:"
}

function _wasp_db {
    _arguments \
        "1: :(migrate-dev studio)"
}
```

**Comment by Martinsos** [comment]
Nice explanation!

I think we should be looking into solution number one -> go with bash autocompletion and just piggy back on it for zsh. As you said, we can ask users what they use, then give them correct instrutcions / do correct actions, and that is it.

Regarding the bash script you wrote -> I would rather we don't have duplicated logic for autocompletion. Right now we actually don't have much logic at all even in bash -> all logic is in Haskell, and bash just proxies it. So we really want sometihng simlar with zhs. Piggybacking on bash sounds like the best trade off between speed and what we get.

**Comment by AmberArr** [comment]
I use a similar script but reusing bash completion functionality.
```zsh
#compdef wasp

local completions
local completion

IFS=$'\n' completions=($( COMP_LINE=$words wasp completion:list ))

for completion in $completions; do
    compadd -- $completion
done
```
But the bash completion now needs some improvement. `wasp new <TAB>` (there is a space after new) just gives me `wasp new new`. 

Switching to `optparse-applicative` might be a option, and we can get [free bash/zsh/fish completion](https://github.com/pcapriotti/optparse-applicative#bash-zsh-and-fish-completions) and no more sync between cli and completion. But the code would look complex IMO.

**Reply by Martinsos** [reply]
Yeah, you are right, I think we should switch to optparse-applicative. It will make code more complex, but I think it is worth it.
